bpo-30971: Improve code readability of json.tool

dhimmel opened the pull request:
Several months ago I worked on pull requests to add new options to the `json.tool` command line utility. These pull requests are still open, pending further design discussion and consensus on whether to add them:

+ https://github.com/python/cpython/pull/345 to add indentation / whitespace options ([`bpo-29636`](http://bugs.python.org/issue29636)).
+ https://github.com/python/cpython/pull/201 to display non-ascii characters ([`bpo-27413`](http://bugs.python.org/issue27413)).

Related, https://github.com/python/cpython/pull/346 fixed a small testing bug in `json.tool` and was merged.

Since #345 and #201 have been inactive for some time, I wanted to take the non-controversial code changes from those PRs. This PR contains readability improvements and minor refactoring to `json.tool`. The goal is to get some of the previous work merged and make the `json.tool` source easier to maintain and enhance going forward. Admittedly, the changes are quite minor.

dhimmel committed:
json.tool: improve code readability

dhimmel commented on Lib/json/tool.py:
```diff
@@ -25,22 +25,22 @@ def main():
     parser.add_argument('outfile', nargs='?', type=argparse.FileType('w'),
                         help='write the output of infile to outfile')
-    parser.add_argument('--sort-keys', action='store_true', default=False,
-                        help='sort the output of dictionaries alphabetically by key')
```
This line exceeded the PEP8 line length, hence the rephrasing.

dhimmel commented on Lib/json/tool.py:
```diff
@@ -25,22 +25,22 @@ def main():
             raise SystemExit(e)
+
+    # Output JSON
+    outfile = options.outfile or sys.stdout
```
Move variable definition close to where it's used.

nirs replied:
Not sure this change helps, it is easier to read when we process all options in one place, and then use the process options.

It would be nice if we could integrate this in options:

    outfile = options.outfile or sys.stdout

Does this work?

     parser.add_argument('outfile', nargs='?', type=argparse.FileType('w'),
                          default=sys.stdout,
                          help='write the output of infile to outfile')

dhimmel commented on Lib/json/tool.py:
```diff
@@ -25,22 +25,22 @@ def main():
     infile = options.infile or sys.stdin
-    outfile = options.outfile or sys.stdout
     sort_keys = options.sort_keys
+    hook = collections.OrderedDict if sort_keys else None
```
Remove the heavily indented if-else for readability

dhimmel commented on Lib/json/tool.py:
```diff
@@ -25,22 +25,22 @@ def main():
                         help='a JSON file to be validated or pretty-printed')
     parser.add_argument('outfile', nargs='?', type=argparse.FileType('w'),
                         help='write the output of infile to outfile')
-    parser.add_argument('--sort-keys', action='store_true', default=False,
```
Remove unnecessary argument.

dhimmel commented on Lib/json/tool.py:
```diff
@@ -25,22 +25,22 @@ def main():
+                        help='sort dictionary outputs alphabetically by key')
     options = parser.parse_args()
 
+    # Read input JSON
```
Comment for readability.

dhimmel reviewed (commented):
I left inline comments for the motivation behind each change.

bitdancer commented:
We generally don't make changes like these unless we are modifying the code for other reasons.

nirs commented on Lib/json/tool.py:
```diff
@@ -25,22 +25,22 @@ def main():
-    parser.add_argument('--sort-keys', action='store_true', default=False,
-                        help='sort the output of dictionaries alphabetically by key')
+    parser.add_argument('--sort-keys', action='store_true',
+                        help='sort dictionary outputs alphabetically by key')
```
... dictionaries output ...?

dhimmel replied:
@nirs I changed it to:

> sort dictionaries output alphabetically by key

as per your suggestion. But would it make more sense as possessive:

> sort dictionaries' output alphabetically by key

nirs replied:
I would check other modules and use the same style.

bitdancer replied:
I would phrase this as "sort alphabetically by key when generating dictionary output".

dhimmel committed:
Address https://github.com/python/cpython/pull/2720#pullrequestreview-50224377

nirs reviewed (approved):
Nice!

dhimmel committed:
Update --sort-keys help message

Use --sort-keys help message based on the json.tool docs at
https://docs.python.org/3.6/library/json.html#basic-usage:

> If sort_keys is true, then the output of dictionaries
> will be sorted by key.

dhimmel commented:
> If this issue is so simple that there’s no need for an issue to track any discussion of what the pull request is trying to solve (e.g. fixing a spelling mistake), then the pull request needs to have the “skip issue” label added to it.

`bedevere/issue-number` is failing. I assumed this PR doesn't require an issue, but let me know if otherwise.

bitdancer commented:
This absolutely needs an issue to discuss whether or not to apply it.  We probably won't, as we don't normally accept "prettify" pull requests.

dhimmel commented:
> This absolutely needs an issue to discuss whether or not to apply it.

Thanks @bitdancer for the guidance. I opened [`bpo-30971`](https://bugs.python.org/issue30971).

dhimmel committed:
Remove comments

Refs https://bugs.python.org/msg298692.
Code was descriptive without the comments.

dhimmel commented:
I'm closing the pull request in response to the advice received in [`bpo-30971`](https://bugs.python.org/issue30971).

Specifically from @bitdancer:

> I think using this PR as a base for your feature PRs, and then committing everything together if they are accepted, would be the way to go.

And from @berkerpeksag:

> Everyone has their own preferences and it's usually not enough to justify code churn.

Will rebase my other PRs with json.tool enhancements on top of these changes.
